packages/focuslock/src/react/index.js

```diff
@@ -0,0 +1,101 @@
+import PropTypes from 'prop-types'
+import React from 'react'
+
+const FOCUSABLE_SELECTORS = [
+  '[contenteditable]:not([contenteditable="false"])',
+  '[tabindex]',
+  'a[href]',
+  'audio[controls]',
+  'button',
+  'iframe',
+  'input',
+  'select',
+  'textarea',
+  'video[controls]'
+]
+
+const hasNegativeTabIndex = el =>
+  el.getAttribute('tabindex') && el.getAttribute('tabindex') < 0
+
+const getFocusableChildNodes = el => {
+  const selectAll = FOCUSABLE_SELECTORS.join(',')
+  const nodelist = el.querySelectorAll(selectAll)
+
+  return Array.from(nodelist || []).filter(node => !hasNegativeTabIndex(node))
+}
+
+class FocusLock extends React.Component {
+  constructor(props) {
+    super(props)
+
+    this.prevFocusedElement = document.activeElement
+
+    this.bindElement = this.bindElement.bind(this)
+    this.handleKeyDown = this.handleKeyDown.bind(this)
+  }
+
+  bindElement(el) {
+    this.el = el
+  }
+
+  componentDidMount() {
+    this.focusableNodes = getFocusableChildNodes(this.el)
+
+    const firstNode = this.focusableNodes[0]
+    if (this.props.autofocus && firstNode) firstNode.focus()
+  }
+
+  componentDidUpdate() {
+    this.focusableNodes = getFocusableChildNodes(this.el)
+  }
+
+  componentWillUnmount() {
+    const { prevFocusedElement: prev } = this
+    if (this.props.returnFocus && prev && prev.focus) prev.focus()
+  }
+
+  handleKeyDown(event) {
+    const isTab = event.key === 'Tab'
+    const withShiftKey = event.shiftKey
+
+    if (!isTab) return
+
+    const { activeElement } = document
+
+    const firstNode = this.focusableNodes[0]
+    const lastNode = this.focusableNodes[this.focusableNodes.length - 1]
+
+    if (activeElement === lastNode) {
+      firstNode.focus()
+      event.preventDefault()
+    } else if (withShiftKey && activeElement === firstNode) {
+      lastNode.focus()
+      event.preventDefault()
+    }
+  }
+
+  render() {
+    const { as: Tag, children } = this.props
+
+    return (
+      <Tag ref={this.bindElement} onKeyDown={this.handleKeyDown}>
+        {children}
+      </Tag>
+    )
+  }
+}
+
+FocusLock.propTypes = {
+  as: PropTypes.oneOf(['div']).isRequired,
```

Comment:
This is a pattern I've seen in Material UI, but we've yet to use it in our components.  

Thus far, when we've wanted to supply semantics, we've passed it as a first child, something like:

```
<FocusLock>
  <div>...
```

I like that approach better in general.  I'd rather have one approach than many.  But I haven't considered long enough whether that approach even works or feels appropriate in this case.  So, we might need another approach.  Or the advantages of this approach, so far unapparent to me, could be enough the we should migrate to it.  What is your take on this subject?

Reply:
i will removed - we add in support for additional tags if needed later